### PR TITLE
Correct initial commitment tx fee affordability checks on open

### PR DESCRIPTION
**Pull request by TheBlueMatt**
Previously, we would reject inbound channels if the funder wasn't
able to meet our channel reserve on their first commitment
transaction only if they also failed to push enough to us for us
to not meet their initial channel reserve as well.

There's not a lot of reason to care about us meeting their reserve,
however - its largely expected that they may not push enough to us
in the initial open to meet it, and its not actually our problem if
they don't.

Further, we used our own fee, instead of the channel's actual fee,
to calculate fee affordability of the initial commitment
transaction.

We resolve both issues here, rewriting the combined affordability
check conditionals in inbound channel open handling and adding a
fee affordability check for outbound channels as well.

The prior code may have allowed a counterparty to start the channel
with "no punishment" states - violating the reason for the reserve
threshold.

This also fixes a few issues in tests that turned up trying to test the fixed logic.

### CODE REVIEW

File: lightning/src/ln/channel.rs
Comment (valentinewallace) [inline lightning/src/ln/channel.rs:675]: This should be able to use one of the `commit_tx_fee` helpers

File: lightning/src/ln/channel.rs
Comment (valentinewallace) [inline lightning/src/ln/channel.rs:935]: Similar, could use one of the `commit_tx_fee` calculation helpers

File: lightning/src/ln/channel.rs
Comment (valentinewallace) [inline lightning/src/ln/channel.rs:937]: Could explicitly give the units as satoshis here

File: lightning/src/ln/channel.rs
Comment (valentinewallace) [inline lightning/src/ln/channel.rs:941]: nit: it's

File: lightning/src/ln/functional_tests.rs
Comment (valentinewallace) [inline lightning/src/ln/functional_tests.rs:1403]: nit: There's a `commit_tx_fee` helper in this file. Similar below

File: lightning/src/ln/channel.rs
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:675]: There is technically no `self` on which to call it currently. I suppose we could make that function take the fee as an argument instead of `&self`? Would you prefer that?

Comment (valentinewallace) [reply lightning/src/ln/channel.rs:675]: Ah. Yeah, I'd prefer that

File: lightning/src/ln/functional_tests.rs
Comment (valentinewallace) [inline lightning/src/ln/functional_tests.rs:620]: nit: accounting

**Review by valentinewallace — APPROVED** [review]

File: lightning/src/ln/functional_tests.rs
Comment (jkczyz) [inline lightning/src/ln/functional_tests.rs:1425]: Consider pulling this from the `TestFeeEstimator` here and elsewhere.

File: lightning/src/ln/channel.rs
Comment (jkczyz) [inline lightning/src/ln/channel.rs:675]: Could you add constant for 4 given it is used in a couple places here and throughout the functional tests?

**Review by jkczyz — APPROVED** [review]

**Comment by TheBlueMatt** [comment]
This is gonna conflict a bit with #1054, so gonna wait for @ariard to confirm which he'd prefer to see land first.

**Comment by TheBlueMatt** [comment]
Squashed without diff:

```
$ git diff-tree -U1 9b7b75de f0cc7882
$
```

**Comment by TheBlueMatt** [comment]
Rebased after #1054 was merged. range diff follows.

```diff
$ git range-diff 2b4ca9e9c53ee676620932754165bcc5e2308153...f0cc7882a3428551c90ec107a3ad112c29673f84 ba50dd57868a78d00ed1a3c7a9089cf39d3ca235...13e4fd586ed0c8439dcada150c4fffa33fa67db9
1:  47d4496fb ! 1:  a33d3b98d Make Channel::commit_tx_fee_msat static and take fee explicitly
    @@ Commit message
      ## lightning/src/ln/channel.rs ##
     @@ lightning/src/ln/channel.rs: impl<Signer: Sign> Channel<Signer> {
      
    -   // Get the fee cost of a commitment tx with a given number of HTLC outputs.
    +   // Get the fee cost in MSATS of a commitment tx with a given number of HTLC outputs.
        // Note that num_htlcs should not include dust HTLCs.
     -  fn commit_tx_fee_msat(&self, num_htlcs: usize) -> u64 {
     +  fn commit_tx_fee_msat(feerate_per_kw: u32, num_htlcs: usize) -> u64 {
    @@ lightning/src/ln/channel.rs: impl<Signer: Sign> Channel<Signer> {
     +          (COMMITMENT_TX_BASE_WEIGHT + num_htlcs as u64 * COMMITMENT_TX_WEIGHT_PER_HTLC) * feerate_per_kw as u64 / 1000 * 1000
        }
      
    -   // Get the commitment tx fee for the local's (i.e. our) next commitment transaction based on the
    +   // Get the fee cost in SATS of a commitment tx with a given number of HTLC outputs.
     @@ lightning/src/ln/channel.rs: impl<Signer: Sign> Channel<Signer> {
                }
      
    @@ lightning/src/ln/channel.rs: impl<Signer: Sign> Channel<Signer> {
                                                && info.next_holder_htlc_id == self.next_holder_htlc_id
                                                && info.next_counterparty_htlc_id == self.next_counterparty_htlc_id
                                                && info.feerate == self.feerate_per_kw {
    --                                                  let actual_fee = self.commit_tx_fee_msat(counterparty_commitment_tx.2);
    -+                                                  let actual_fee = Self::commit_tx_fee_msat(self.feerate_per_kw, counterparty_commitment_tx.2);
    +-                                                  let actual_fee = self.commit_tx_fee_msat(commitment_stats.num_nondust_htlcs);
    ++                                                  let actual_fee = Self::commit_tx_fee_msat(self.feerate_per_kw, commitment_stats.num_nondust_htlcs);
                                                        assert_eq!(actual_fee, info.fee);
                                                }
                                }
2:  28098fc81 ! 2:  1d30e0689 Rewrite test_update_fee_that_funder_cannot_afford to avoid magic
    @@ Commit message
         and so that it is easier to re-derive constants in the future as
         needed.
     
    + ## lightning/src/ln/channel.rs ##
    +@@ lightning/src/ln/channel.rs: pub(crate) const FUNDING_CONF_DEADLINE_BLOCKS: u32 = 2016;
    + /// size 2. However, if the number of concurrent update_add_htlc is higher, this still
    + /// leads to a channel force-close. Ultimately, this is an issue coming from the
    + /// design of LN state machines, allowing asynchronous updates.
    +-const CONCURRENT_INBOUND_HTLC_FEE_BUFFER: u32 = 2;
    ++pub(crate) const CONCURRENT_INBOUND_HTLC_FEE_BUFFER: u32 = 2;
    + 
    + // TODO: We should refactor this to be an Inbound/OutboundChannel until initial setup handshaking
    + // has been completed, and then turn into a Channel to get compiler-time enforcement of things like
    +
      ## lightning/src/ln/functional_tests.rs ##
    +@@ lightning/src/ln/functional_tests.rs: use chain::channelmonitor::{ChannelMonitor, CLTV_CLAIM_BUFFER, LATENCY_GRACE_PER
    + use chain::transaction::OutPoint;
    + use chain::keysinterface::BaseSign;
    + use ln::{PaymentPreimage, PaymentSecret, PaymentHash};
    +-use ln::channel::{COMMITMENT_TX_BASE_WEIGHT, COMMITMENT_TX_WEIGHT_PER_HTLC};
    ++use ln::channel::{COMMITMENT_TX_BASE_WEIGHT, COMMITMENT_TX_WEIGHT_PER_HTLC, CONCURRENT_INBOUND_HTLC_FEE_BUFFER};
    + use ln::channelmanager::{ChannelManager, ChannelManagerReadArgs, PaymentId, RAACommitmentOrder, PaymentSendFailure, BREAKDOWN_TIMEOUT, MIN_CLTV_EXPIRY_DELTA};
    + use ln::channel::{Channel, ChannelError};
    + use ln::{chan_utils, onion_utils};
     @@ lightning/src/ln/functional_tests.rs: fn test_update_fee_that_funder_cannot_afford() {
        let node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
        let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
        let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
    --  let channel_value = 1888;
    +-  let channel_value = 1977;
     -  let chan = create_announced_chan_between_nodes_with_value(&nodes, 0, 1, channel_value, 700000, InitFeatures::known(), InitFeatures::known());
     +  let channel_value = 5000;
     +  let push_sats = 700;
     +  let chan = create_announced_chan_between_nodes_with_value(&nodes, 0, 1, channel_value, push_sats * 1000, InitFeatures::known(), InitFeatures::known());
        let channel_id = chan.2;
    -+  let bs_channel_reserve_sats = Channel::<EnforcingSigner>::get_holder_selected_channel_reserve_satoshis(channel_value);
    - 
    +   let secp_ctx = Secp256k1::new();
    +-
     -  let feerate = 260;
    -+  // Calculate the maximum feerate that A can afford:
    -+  let feerate = ((channel_value - bs_channel_reserve_sats - push_sats) * 1000 / COMMITMENT_TX_BASE_WEIGHT) as u32;
    ++  let bs_channel_reserve_sats = Channel::<EnforcingSigner>::get_holder_selected_channel_reserve_satoshis(channel_value);
    ++
    ++  // Calculate the maximum feerate that A can afford. Note that we don't send an update_fee
    ++  // CONCURRENT_INBOUND_HTLC_FEE_BUFFER HTLCs before actually running out of local balance, so we
    ++  // calculate two different feerates here - the expected local limit as well as the expected
    ++  // remote limit.
    ++  let feerate = ((channel_value - bs_channel_reserve_sats - push_sats) * 1000 / (COMMITMENT_TX_BASE_WEIGHT + CONCURRENT_INBOUND_HTLC_FEE_BUFFER as u64 * COMMITMENT_TX_WEIGHT_PER_HTLC)) as u32;
    ++  let non_buffer_feerate = ((channel_value - bs_channel_reserve_sats - push_sats) * 1000 / COMMITMENT_TX_BASE_WEIGHT) as u32;
        {
                let mut feerate_lock = chanmon_cfgs[0].fee_estimator.sat_per_kw.lock().unwrap();
                *feerate_lock = feerate;
    @@ lightning/src/ln/functional_tests.rs: fn test_update_fee_that_funder_cannot_affo
     +          *feerate_lock += 4;
        }
        nodes[0].node.timer_tick_occurred();
    -   check_added_monitors!(nodes[0], 1);
    +-  nodes[0].logger.assert_log("lightning::ln::channel".to_string(), format!("Cannot afford to send new feerate at {}", feerate + 2), 1);
    ++  nodes[0].logger.assert_log("lightning::ln::channel".to_string(), format!("Cannot afford to send new feerate at {}", feerate + 4), 1);
    +   check_added_monitors!(nodes[0], 0);
    + 
    +   const INITIAL_COMMITMENT_NUMBER: u64 = 281474976710654;
    +@@ lightning/src/ln/functional_tests.rs: fn test_update_fee_that_funder_cannot_afford() {
    +           let mut htlcs: Vec<(HTLCOutputInCommitment, ())> = vec![];
    +           let commitment_tx = CommitmentTransaction::new_with_auxiliary_htlc_data(
    +                   INITIAL_COMMITMENT_NUMBER - 1,
    +-                  700,
    +-                  999,
    ++                  push_sats,
    ++                  channel_value - push_sats - commit_tx_fee_msat(non_buffer_feerate + 4, 0) / 1000,
    +                   false, local_funding, remote_funding,
    +                   commit_tx_keys.clone(),
    +-                  feerate + 124,
    ++                  non_buffer_feerate + 4,
    +                   &mut htlcs,
    +                   &local_chan.channel_transaction_parameters.as_counterparty_broadcastable()
    +           );
    +@@ lightning/src/ln/functional_tests.rs: fn test_update_fee_that_funder_cannot_afford() {
    + 
    +   let update_fee = msgs::UpdateFee {
    +           channel_id: chan.2,
    +-          feerate_per_kw: feerate + 124,
    ++          feerate_per_kw: non_buffer_feerate + 4,
    +   };
    + 
    +   nodes[1].node.handle_update_fee(&nodes[0].node.get_our_node_id(), &update_fee);
3:  3f82a795f ! 3:  940ef0537 Correct initial commitment tx fee affordability checks on open
    @@ fuzz/src/full_stack.rs: mod tests {
                assert_eq!(log_entries.get(&("lightning::ln::peer_handler".to_string(), "Handling SendAcceptChannel event in peer_handler for node 030000000000000000000000000000000000000000000000000000000000000002 for channel ff4f00f805273c1b203bb5ebf8436bfde57b3be8c2f5e95d9491dbb181909679".to_string())), Some(&1)); // 1
     
      ## lightning/src/ln/channel.rs ##
    -@@ lightning/src/ln/channel.rs: const FEE_SPIKE_BUFFER_FEE_INCREASE_MULTIPLE: u64 = 2;
    - /// This constant is the one suggested in BOLT 2.
    - pub(crate) const FUNDING_CONF_DEADLINE_BLOCKS: u32 = 2016;
    +@@ lightning/src/ln/channel.rs: pub(crate) const FUNDING_CONF_DEADLINE_BLOCKS: u32 = 2016;
    + /// design of LN state machines, allowing asynchronous updates.
    + pub(crate) const CONCURRENT_INBOUND_HTLC_FEE_BUFFER: u32 = 2;
      
     +/// When a channel is opened, we check that the funding amount is enough to pay for relevant
     +/// commitment transaction fees, with at least this many HTLCs present on the commitment
    @@ lightning/src/ln/functional_tests.rs: use chain::channelmonitor::{ChannelMonitor
      use chain::transaction::OutPoint;
      use chain::keysinterface::BaseSign;
      use ln::{PaymentPreimage, PaymentSecret, PaymentHash};
    --use ln::channel::{COMMITMENT_TX_BASE_WEIGHT, COMMITMENT_TX_WEIGHT_PER_HTLC};
    -+use ln::channel::{COMMITMENT_TX_BASE_WEIGHT, COMMITMENT_TX_WEIGHT_PER_HTLC, MIN_AFFORDABLE_HTLC_COUNT};
    +-use ln::channel::{COMMITMENT_TX_BASE_WEIGHT, COMMITMENT_TX_WEIGHT_PER_HTLC, CONCURRENT_INBOUND_HTLC_FEE_BUFFER};
    ++use ln::channel::{COMMITMENT_TX_BASE_WEIGHT, COMMITMENT_TX_WEIGHT_PER_HTLC, CONCURRENT_INBOUND_HTLC_FEE_BUFFER, MIN_AFFORDABLE_HTLC_COUNT};
      use ln::channelmanager::{ChannelManager, ChannelManagerReadArgs, PaymentId, RAACommitmentOrder, PaymentSendFailure, BREAKDOWN_TIMEOUT, MIN_CLTV_EXPIRY_DELTA};
      use ln::channel::{Channel, ChannelError};
      use ln::{chan_utils, onion_utils};
4:  f0cc7882a = 4:  13e4fd586 Test fixed channel reserve checks on channel open
$
```

**Review by valentinewallace — APPROVED** [review]

**Review by jkczyz — APPROVED** [review]

**Pull request merged**
